swagger_zipkin/otel_decorator.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+import importlib
+from contextlib import contextmanager
+from typing import Any
+from typing import TYPE_CHECKING
+from typing import TypeVar
+
+from opentelemetry import trace
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+from opentelemetry.trace.span import format_span_id
+from opentelemetry.trace.span import format_trace_id
+from opentelemetry.trace.span import TraceFlags
+from typing_extensions import ParamSpec
+
+from swagger_zipkin.decorate_client import Client
+from swagger_zipkin.decorate_client import decorate_client
+from swagger_zipkin.decorate_client import Resource
+
+if TYPE_CHECKING:
+    import pyramid.request.Request   # type: ignore # noqa: F401
+
+T = TypeVar('T', covariant=True)
+P = ParamSpec('P')
+
+tracer = trace.get_tracer("otel_decorator")
+
+
+class OtelResourceDecorator:
+    """A wrapper to the swagger resource.
+
+    :param resource: A resource object. eg. `client.pet`, `client.store`.
+    :type resource: :class:`swaggerpy.client.Resource` or :class:`bravado_core.resource.Resource`
+    """
+
+    def __init__(self, resource: Client, client_identifier: str, smartstack_namespace: str) -> None:
+        self.resource = resource
+        self.client_identifier = client_identifier
+        self.smartstack_namespace = smartstack_namespace
+
+    def __getattr__(self, name: str) -> Resource:
+        return decorate_client(self.resource, self.with_headers, name)
+
+    def with_headers(self, call_name: str, *args: Any, **kwargs: Any) -> Any:
+        kwargs.setdefault('_request_options', {})
+        request_options: dict = kwargs['_request_options']
+        request_options.setdefault('headers', {})
+
+        # what is the right way to get the Request object. can we use contruct_request
+        # https://github.com/Yelp/bravado/blob/master/bravado/client.py#L283C5-L283C22
```

Comment:
What is the right way to get the Request object?

Reply:
tl;dr
I've used contruct_request from bravado to get the Request object.

Yeah that makes sense this is instrumenting for a outgoing call (not an incoming call where pyramid.request.Request would be available).
For zipkin_decorator it was straightforward since we the zipkin context was used to just inject B3 headers in the outgoing call.
For otel, the context itself is available in the trace object and we can inject zipkin and otel headers. But the additional functionality being added here is the generation of a CLIENT span. For this purpose a bunch of fields are needed: url, path, etc and hence the need for a Request object. This wont be available in the otel/zipkin context.

Reply:
Ah, I see I misunderstood which request you meant. I think this should be fine, although I don't think any other bravado decorator has done anything like this before.